Lib/email/_header_value_parser.py

```diff
@@ -2726,12 +2726,23 @@ def _fold_as_ew(to_encode, lines, maxlen, last_ew, ew_combine_allowed, charset):
             continue
         first_part = to_encode[:text_space]
         ew = _ew.encode(first_part, charset=encode_as)
-        excess = len(ew) - remaining_space
-        if excess > 0:
-            # encode always chooses the shortest encoding, so this
-            # is guaranteed to fit at this point.
-            first_part = first_part[:-excess]
-            ew = _ew.encode(first_part)
+        if len(ew) > remaining_space:
+            # Find the longest first_part
+            # since len(_ew.encode(to_encode[:x])) is a non-linear
+            # monotonically increasing function, and calculating the
```

Comment:
`_ew.encode` is [biased towards the 'q' encoding][encode]. This might violate the assumption of a monotonically increasing function for some corner cases. (This was already the case for the old code.)

I hope to find the time to write a test case for this.

 [encode]: https://github.com/python/cpython/blob/c3f55be7dd012b7e92901627d0b31c21e983ccb4/Lib/email/_encoded_words.py#L228